pkg/graveler/retention/garbage_collection_manager.go

```diff
@@ -108,15 +112,16 @@ func (m *GarbageCollectionManager) GetRunExpiredCommits(ctx context.Context, sto
 }
 
 func (m *GarbageCollectionManager) SaveGarbageCollectionCommits(ctx context.Context, storageNamespace graveler.StorageNamespace, repositoryID graveler.RepositoryID, rules *graveler.GarbageCollectionRules, previouslyExpiredCommits []graveler.CommitID) (string, error) {
-	branchIterator, err := m.refManager.ListBranches(ctx, repositoryID)
-	if err != nil {
-		return "", fmt.Errorf("list repository branches: %w", err)
-	}
 	commitGetter := &RepositoryCommitGetter{
 		refManager:   m.refManager,
 		repositoryID: repositoryID,
 	}
-	gcCommits, err := GetGarbageCollectionCommits(ctx, branchIterator, commitGetter, rules, previouslyExpiredCommits)
+	branchIterator := ref.NewBranchIterator(ctx, m.db, repositoryID, 1000, ref.WithOrderBy("commit_id"))
+	// get all commits that are not the first parent of any commit:
+	mainAncestryLeaves := "NOT EXISTS (SELECT * FROM graveler_commits c2 WHERE c2.repository_id=graveler_commits.repository_id AND c2.parents[1]=graveler_commits.id"
```

Comment:
This looks like something that would be quite hard for the query planner!  Perhaps we should reduce the isolation level, and/or fix it as  a proper anti-join.

Reply:
This is the postgres way to anti-join :|